text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,658 @@
+ as arguments. This leads to method signatures being long and needlessly unwieldy.
+
+Note: We had initially intended to overload all of these classes method signatures in 
+this way, to ensure that `RecordData` implementations could be `singleton`s, but we failed
+to correctly implement the `RecordData` RFC in this regard and a near future RFC will look
+at rectifying this using the `Identity` APIs introduced here.
+
+Moving to a unified concept of `identity` opens up a path to clean up these method
+ signatures.
+
+### Operations
+
+Operations are a foundational concept for `acid` transactions. Without the ability to 
+ describe an operation and a clear mental model of what operations exist and achieve, it
+ is difficult to understand how an action affects state. Granularity and clarity is key.
+
+While `ember-data` does not yet have concepts of operations or transactions, mutations and
+ updates are applied directly, there are many areas we could improve upon by introducing
+them. As with data, operations upon data should be serializable so that local state can 
+be accurately cached on local clients.
+
+### Nested Saves / API Transactions / Websocket Support
+
+Many applications wish to create or update and save multiple records together. Achieving
+ this in `ember-data` today is unweildy and has many difficult edge cases: one of which
+ is correctly matching data received back from the API to the newly created records already
+ on the client.
+
+A similar edge case occurs when a newly created record is saved for the first time and 
+ prior to receiving the request response the same record is recieved via another means
+(background polling, websocket subscription, etc.). We have no means of matching the record
+returned by the alternative means to that of the request, leading to a second cache entry 
+being created and an error once the initial request completes.
+
+One solution has been to generate and assign `id`s for records on the client, but this is
+ not always desireable. These scenarios are a major motivation for `lid` in the `json-api`
+spec. Users wishing to solve these cases would be able to serialize the `lid` of the 
+`Identifier` for a newly created record and reflect that `lid` back in any payloads send 
+from their API for the session to correctly match the payloads to the record.
+
+### Better Cache Serialization & Improved Infra for Offline Support
+
+In order to enable users to achieve full offline support, or to serialize the store 
+ or transport across the wire (for example as an advanced  fastboot rehydration mode)
+the entire state of the store needs to be serializable. This RFC introduces the 
+foundation for mechanisms through which this can be later achieved.
+
+----
+
+## Detailed design
+
+```typescript
+export interface Identifier {
+  lid: string;
+}
+
+export interface RecordIdentifier extends Identifier {
+  id: string | null;
```

Comment:
I believe this is more properly `id?: string` since `id` may be completely missing (which seems preferable to `null`).

Reply:
`null` is correct as that is what new records will have; however, a missing id when `lid` is available is allowable, so updated to allow not having `id` in this case.